### PR TITLE
Add .WithDefaultRedirectUri() method to the PublicClientApplicationBuilder

### PR DESCRIPTION
**Pull request by trwalke**
implementation of .WithDefaultRedirectUri() [#1186](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1186)

### CODE REVIEW

File: src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs [hunk 1]
```diff
@@ -56,6 +56,7 @@ public string ClientVersion
         public LogCallback LoggingCallback { get; internal set; }
         public string Component { get; internal set; }
         public IDictionary<string, string> ExtraQueryParameters { get; internal set; } = new Dictionary<string, string>();
+        public bool UseNewDefaultRedirectUri { get; internal set; }
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs]: "New" is an overloaded word. It might be "new" today, but 6 months from now, we will have forgotten all about the "old" obo:urn...

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 2]
```diff
@@ -50,6 +50,21 @@ internal PublicClientApplicationBuilder WithUserTokenLegacyCachePersistenceForTe
             return this;
         }
 
+        /// <summary>
+        /// Sets the public client application to use the latest default redirect URI as opposed to urn:ietf:wg:oauth:2.0:oob.
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: Needs review, I don't think users need to understand "latest" and the fact that in the past we were using urn:ieft...

I would reword it like "Configures the pca to use the recommended redirect Uri for the platform. If you are calling your API from a .NET Classic app - https://login.microsoftonline.com/common/oauth2/nativeclient, ....."

File: src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs (hunk 1 above)
Comment (bgavrilMS) [reply src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs]: UseRecommendedDefaultRedirectUri ?

File: src/Microsoft.Identity.Client/Platforms/Android/AndroidPlatformProxy.cs [hunk 3]
```diff
@@ -89,7 +89,7 @@ public override string GetBrokerOrRedirectUri(Uri redirectUri)
         }
 
         /// <inheritdoc />
-        public override string GetDefaultRedirectUri(string clientId)
+        public override string GetDefaultRedirectUri(string clientId, bool useNewRedirectURI = false)
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Client/Platforms/Android/AndroidPlatformProxy.cs]: useRecommended..

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs [hunk 4]
```diff
@@ -326,6 +326,47 @@ public void AcquireTokenSameUserTest()
...
+                                                                            .WithDefaultRedirectUri()
+                                                                            .BuildConcrete();
+
+                //Validate new default redirect uri
+#if DESKTOP
+                Assert.AreEqual(app.AppConfig.RedirectUri, "https://login.microsoftonline.com/common/oauth2/nativeclient");
+#elif NET_CORE
+                Assert.AreEqual(app.AppConfig.RedirectUri, "https://localhost");
+#endif
+                MsalMockHelpers.ConfigureMockWebUI(
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: I'd remove this second part of the unit test test and modify some of the integration tests instead.

File: src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs (hunk 1 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs]: Shouldn't this just be UseDefaultRedirectUri?  This is now the default.  If we have a setting for using the old one (to force it for some reason) we should have UseOobRedirectUri or something./

**Comment by MarkZuber** [comment]
Need to update our build rules.  This is running MSAL V3 PR/automation and MSAL V4 PR/automation.

File: src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs (hunk 1 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs]: I believe we are still going to keep "urn:ietf:wg:oauth:2.0:oob" as the default redirect uri to avoid breaking customers which is why I named this "new" instead of just "UseDefaultRedirectUri".

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs (hunk 4 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: makes sense to remove the second part however I dont understand what you mean by "modify some of the integration tests instead". Are you saying that I need to add this check to some integration tests? Are unit tests not sufficient to test this scenario?

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs (hunk 2 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: @jmprieur any thoughts on what this comment should be?

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 5]
```diff
@@ -50,6 +50,21 @@ internal PublicClientApplicationBuilder WithUserTokenLegacyCachePersistenceForTe
             return this;
         }
 
+        /// <summary>
+        /// Configures the pca to use the recommended redirect Uri for the platform.
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: pca => Public client application.
redirect Uri => reply URI (to be aligned with the portal)

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 6]
```diff
@@ -50,6 +50,21 @@ internal PublicClientApplicationBuilder WithUserTokenLegacyCachePersistenceForTe
             return this;
         }
 
+        /// <summary>
+        /// Configures the pca to use the recommended redirect Uri for the platform. 
+        /// If you are calling your API from a 
+        /// .NET Classic app - https://login.microsoftonline.com/common/oauth2/nativeclient
+        /// UWP - value of WebAuthenticationBroker.GetCurrentApplicationCallbackUri()
+        /// .NET Core - https://localhost
+        /// </summary>
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: @jennyf19 : not for now, but when we support brokers, will it make sense to also have this method set the URI used for the broker scenario? (thinking ahead here)

File: src/Microsoft.Identity.Client/Platforms/Mac/MacPlatformProxy.cs [hunk 7]
```diff
@@ -78,7 +78,7 @@ public override string GetBrokerOrRedirectUri(Uri redirectUri)
         }
 
         /// <inheritdoc />
-        public override string GetDefaultRedirectUri(string clientId)
+        public override string GetDefaultRedirectUri(string clientId, bool useRecommendedRedirectURI = false)
         {
             return Constants.DefaultRedirectUri;
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Platforms/Mac/MacPlatformProxy.cs]: I think we should use the same here as for desktop

File: src/Microsoft.Identity.Client/Platforms/netstandard13/NetStandard13PlatformProxy.cs [hunk 8]
```diff
@@ -43,7 +43,7 @@ public override string GetBrokerOrRedirectUri(Uri redirectUri)
         }
 
         /// <inheritdoc />
-        public override string GetDefaultRedirectUri(string clientId)
+        public override string GetDefaultRedirectUri(string clientId, bool useRecommendedRedirectURI = false)
         {
             return Constants.DefaultRedirectUri;
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Platforms/netstandard13/NetStandard13PlatformProxy.cs]: We should use the same as for desktop imo

File: src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs [hunk 9]
```diff
@@ -77,8 +77,13 @@ public override string GetBrokerOrRedirectUri(Uri redirectUri)
         }
 
         /// <inheritdoc />
-        public override string GetDefaultRedirectUri(string clientId)
+        public override string GetDefaultRedirectUri(string clientId, bool useRecommendedRedirectURI = false)
         {
+            if (useRecommendedRedirectURI)
+            {
+                return Constants.DefaultNetCoreRedirectUri;
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs]: Why is this the default for netcore?

File: src/Microsoft.Identity.Client/Core/Constants.cs [hunk 10]
```diff
@@ -14,6 +14,8 @@ internal static class Constants
 
         public const string UapWEBRedirectUri = "https://sso"; // only ADAL supports WEB
         public const string DefaultRedirectUri = "urn:ietf:wg:oauth:2.0:oob";
+        public const string DefaultDesktopRedirectUri = "https://login.microsoftonline.com/common/oauth2/nativeclient";
+        public const string DefaultNetCoreRedirectUri = "https://localhost";
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Core/Constants.cs]: >DefaultNetCoreRedirectUri [](start = 28, length = 25)

Why this? Why can't this be the same as the one above?

File: src/Microsoft.Identity.Client/Platforms/Android/AndroidPlatformProxy.cs [hunk 11]
```diff
@@ -89,7 +89,7 @@ public override string GetBrokerOrRedirectUri(Uri redirectUri)
         }
 
         /// <inheritdoc />
-        public override string GetDefaultRedirectUri(string clientId)
+        public override string GetDefaultRedirectUri(string clientId, bool useRecommendedRedirectURI = false)
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client/Platforms/Android/AndroidPlatformProxy.cs]: nit:  .net naming conventions:  useRecommendedRedirectUri

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs [hunk 12]
```diff
@@ -326,6 +326,34 @@ public void AcquireTokenSameUserTest()
...
+                                                                            .WithHttpManager(harness.HttpManager)
+                                                                            .WithTelemetry(new TraceTelemetryConfig())
+                                                                            .WithDefaultRedirectUri()
+                                                                            .BuildConcrete();
+
+                //Validate new default redirect uri
+#if DESKTOP
+                Assert.AreEqual(app.AppConfig.RedirectUri, "https://login.microsoftonline.com/common/oauth2/nativeclient");
+#elif NET_CORE
+                Assert.AreEqual(app.AppConfig.RedirectUri, "https://localhost");
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: >Assert.AreEqual(app.AppConfig.RedirectUri, "https://localhost"); [](start = 16, length = 64)

I'm not convinced this is right.

File: src/Microsoft.Identity.Client/Core/Constants.cs (hunk 10 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Core/Constants.cs]: not sure but it is according to the spec outlined here
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1186

File: src/Microsoft.Identity.Client/Platforms/net45/NetDesktopPlatformProxy.cs [hunk 13]
```diff
@@ -146,8 +146,13 @@ public override string GetBrokerOrRedirectUri(Uri redirectUri)
         }
 
         /// <inheritdoc />
-        public override string GetDefaultRedirectUri(string clientId)
+        public override string GetDefaultRedirectUri(string clientId, bool useRecommendedRedirectURI = false)
         {
+            if (useRecommendedRedirectURI)
+            {
+                return Constants.DefaultDesktopRedirectUri;
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client/Platforms/net45/NetDesktopPlatformProxy.cs]: Is this not relevant for macOS which is also a desktop platform?

File: src/Microsoft.Identity.Client/Core/Constants.cs (hunk 10 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/Core/Constants.cs]: Perhaps we need to have a WithLocalRedirectUri

---
In reply to: [297898748](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1231#discussion_r297898748) [](ancestors = 297898748)

Comment (henrik-me) [reply src/Microsoft.Identity.Client/Core/Constants.cs]: @JM ... not sure we should do this. 

---
In reply to: [297899005](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1231#discussion_r297899005) [](ancestors = 297899005)

File: src/Microsoft.Identity.Client/Platforms/net45/NetDesktopPlatformProxy.cs (hunk 13 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Platforms/net45/NetDesktopPlatformProxy.cs]: added it

File: src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs (hunk 9 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs]: So that the system web view can be used: See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/msal-net-4#net-core-now-supports-interactive-authentication

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs (hunk 12 above)
Comment (jmprieur) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: @henrik-me would you mean that we should not propose localhost for .NET Core? or do you also question UWP?

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 14]
```diff
@@ -50,6 +50,36 @@ internal PublicClientApplicationBuilder WithUserTokenLegacyCachePersistenceForTe
...
+        /// <term>.NET desktop</term>
+        /// <term><c>https://login.microsoftonline.com/common/oauth2/nativeclient</c></term>
+        /// </item>
+        /// <item>
+        /// <term>UWP</term>
+        /// <term>value of <c>WebAuthenticationBroker.GetCurrentApplicationCallbackUri()</c></term>
+        /// </item>
+        /// <item>
+        /// <term>.NET Core</term>
+        /// <term><c>https://localhost</c></term>
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: It's http, not https. Needs an E2E test :)

Comment (jmprieur) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: my bad. and yes, it needs en E2E test.

Comment (trwalke) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: This is easy to E2E test on desktop and net core but we dont have a way of testing this on UWP

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs (hunk 12 above)
Comment (henrik-me) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: At this time I'm talking about netcore. UWP I believe has a different schema?

File: src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs (hunk 9 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs]: I'm not a fan of combining. it's also possible to use the system browser on desktop. IMO, we should have another URI, WithLocalhostRedirectUri or we should specify it by default when using System Browsers on netcore and desktop... 

---
In reply to: [298076066](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1231#discussion_r298076066) [](ancestors = 298076066)

File: src/Microsoft.Identity.Client/Platforms/net45/NetDesktopPlatformProxy.cs [hunk 15]
```diff
@@ -146,8 +146,13 @@ public override string GetBrokerOrRedirectUri(Uri redirectUri)
         }
 
         /// <inheritdoc />
-        public override string GetDefaultRedirectUri(string clientId)
+        public override string GetDefaultRedirectUri(string clientId, bool useRecommendedRedirectUri = false)
         {
+            if (useRecommendedRedirectUri)
+            {
+                return Constants.DefaultDesktopRedirectUri;
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Platforms/net45/NetDesktopPlatformProxy.cs]: >Constants.DefaultDesktopRedirectUri; [](start = 23, length = 36)

Perhaps this for non system browser and for system browser it should be localhost?

File: src/Microsoft.Identity.Client/Core/Constants.cs (hunk 10 above)
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Core/Constants.cs]: Should probably not be named netcore. Should be named LocalhostRedirectUri imo. And the one above should be renamed to NativeClientRedirectUri

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 16]
```diff
@@ -50,6 +50,36 @@ internal PublicClientApplicationBuilder WithUserTokenLegacyCachePersistenceForTe
...
+        /// <item>
+        /// <term>.NET desktop</term>
+        /// <term><c>https://login.microsoftonline.com/common/oauth2/nativeclient</c></term>
+        /// </item>
+        /// <item>
+        /// <term>UWP</term>
+        /// <term>value of <c>WebAuthenticationBroker.GetCurrentApplicationCallbackUri()</c></term>
+        /// </item>
+        /// <item>
+        /// <term>.NET Core</term>
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: ><term>.NET Core</term> [](start = 12, length = 22)

Suggestion: Change this to be: For system browser on netcore and net desktop

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs [hunk 17]
```diff
@@ -50,6 +50,36 @@ internal PublicClientApplicationBuilder WithUserTokenLegacyCachePersistenceForTe
...
+        /// <term><c>https://localhost</c></term>
+        /// </item>
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: >// </item> [](start = 9, length = 10)

If we are mentioning specifics for netcore and netdesktop perhaps we should also mention specifics on the other platforms? @JM?

File: src/Microsoft.Identity.Client/Core/Constants.cs (hunk 10 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Core/Constants.cs]: Will be address in a later PR

File: src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs (hunk 9 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Platforms/netcore/NetCorePlatformProxy.cs]: Will be address in a later PR

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs (hunk 12 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: Will be address in a later PR

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs (hunk 14 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: Will be address in a later PR

File: src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs (hunk 17 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs]: Will be address in a later PR
